### PR TITLE
ROX-24281: do not create matcher client if undesired

### PR DESCRIPTION
**Pull request by RTann**
## Description

The current implementation meant Sensors which used Scanner V4 would continuously print debug logs indicating it is unable to connect to the Scanner V4 Matcher. It also wastes resources attempting to connect to something which we know does not exist. I expect this same thing to happen to Scanner V4 Matcher, as it connects to the Scanner V4 Indexer, which does not implement the Matcher service.

This PR adds an option to the Scanner V4 client to make it clear when only indexing is desired so we do not waste any resources/logging.

## Checklist
- [x] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

Tried adding unit tests, but it was proving to be a huge pain (unless I was just doing it wrong). The changes are pretty trivial, though I still tested it with a cluster to confirm it works.

### Here I tell how I validated my change

* Deployed ACS Central + Secured Cluster 4.4.2 to OCP via Helm
  * Debug logs were enabled
* Scan image:
  * `roxctl -e <central endpoint> [--insecure-skip-tls-verify] scan image -i quay.io/stackrox-io/scanner:4.4.2`
* Inspect Sensor logs to find the debug log exists
  * `grpc_internal: 2024/06/04 18:30:06.913416 logging.go:59: Debug: [core][Channel #26 SubChannel #27] grpc: addrConn.createTransport failed to connect to {Addr: "[::1]:8443", ServerName: "scanner-v4-matcher.stackrox.svc", }. Err: connection error: desc = "transport: authentication handshake failed: x509: certificate is valid for sensor.stackrox, sensor.stackrox.svc, sensor-webhook.stackrox.svc, not scanner-v4-matcher.stackrox.svc"`
* Redeploy Secured Cluster with version 4.4.x-848-g11a6c13d73 to OCP via Helm
  * Debug logs were enabled
* Scan image:
  * `roxctl -e <central endpoint> [--insecure-skip-tls-verify] scan image -i quay.io/stackrox-io/scanner:4.4.2`
* Inspect Sensor logs and find `scanner-v4-matcher` is no longer referenced

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by vikin91 — APPROVED** [review]

File: sensor/common/scannerclient/grpc_client.go [hunk 1]
```diff
@@ -164,7 +164,7 @@ func dialV2() (ScannerClient, error) {
 // dialV4 connect to scanner V4 gRPC and return a new ScannerClient.
 func dialV4() (ScannerClient, error) {
 	ctx := context.Background()
-	c, err := client.NewGRPCScanner(ctx, client.WithIndexerAddress(env.ScannerV4IndexerEndpoint.Setting()))
+	c, err := client.NewGRPCScanner(ctx, client.IndexerOnly(), client.WithIndexerAddress(env.ScannerV4IndexerEndpoint.Setting()))
```
Comment (jvdm) [inline sensor/common/scannerclient/grpc_client.go:167]: Why don't we use `WithAddress(address)` instead of `client.WithIndexerAddress(address)`?

Comment (RTann) [reply sensor/common/scannerclient/grpc_client.go:167]: For this particular use case, it doesn't really matter. Two reasons why I went it though:

1. It was already there, so the changes are minimized
2. Makes it clear our main focus here is the Indexer and not also the Matcher

**Review by dcaravel — APPROVED** [review]
LGTM

File: sensor/common/scannerclient/grpc_client.go (hunk 1 above)
Comment (jvdm) [reply sensor/common/scannerclient/grpc_client.go:167]: This was changed from `WithAddress` at #9555 for no apparent reason. It looks like it was a misunderstanding of the client's design. The options for setting addresses are not related to the mode of the Scanner we're connecting to. The client was meant to connect to either one instance (combo) or two separate instances (split). If an API is not supported, the server should return "UnsupportedOperation.".

What was the reason for the move to `WithIndexerAddress()`? If none, let's go back to `WithAddress()`.

Comment (jvdm) [reply sensor/common/scannerclient/grpc_client.go:167]: This was changed from `WithAddress` at #9555 for no apparent reason. It looks like it was a misunderstanding of the client's design. The options for setting addresses are not related to the mode of the Scanner we're connecting to. The client was meant to connect to either one instance (combo) or two separate instances (split). If an API is not supported, the server should return "UnsupportedOperation.".

What was the reason for the move to `WithIndexerAddress()`? If none, let's go back to `WithAddress()`.

Comment (RTann) [reply sensor/common/scannerclient/grpc_client.go]: [2da93e6](https://github.com/stackrox/stackrox/pull/11161/commits/2da93e6646395df9c3e4cf2f5b69066b161503e6)

File: sensor/common/scannerclient/grpc_client.go [hunk 2]
```diff
@@ -164,7 +164,7 @@ func dialV2() (ScannerClient, error) {
 // dialV4 connect to scanner V4 gRPC and return a new ScannerClient.
 func dialV4() (ScannerClient, error) {
 	ctx := context.Background()
-	c, err := client.NewGRPCScanner(ctx, client.WithIndexerAddress(env.ScannerV4IndexerEndpoint.Setting()))
+	c, err := client.NewGRPCScanner(ctx, client.IndexerOnly(), client.WithAddress(env.ScannerV4IndexerEndpoint.Setting()))
```
Comment (jvdm) [inline sensor/common/scannerclient/grpc_client.go:167]: I see you moved away from `WithIndexerAddress()`. Why do we need `IndexerOnly()` now?

Comment (RTann) [reply sensor/common/scannerclient/grpc_client.go:167]: ah, I see your point now

Comment (RTann) [reply sensor/common/scannerclient/grpc_client.go]: Updated to remove `IndexerOnly`, and I updated the comments for `WithIndexerAddress` and `WithMatcherAddress` to make it clear when to use those

File: pkg/scannerv4/client/options.go [hunk 3]
```diff
@@ -98,6 +98,9 @@ func SkipIndexerTLSVerification(o *options) {
 }
 
 // WithIndexerAddress specifies the gRPC address to connect.
+// This should be used when the client wants to reach out to both the Indexer and Matcher,
+// but they live at different addresses. When the client only wants to reach the Indexer or
+// both the Indexer and Matcher are at the same address, then use WithAddress.
```
Comment (jvdm) [inline pkg/scannerv4/client/options.go:103]: Nitpick: A better explanation of this option is:

```suggestion
// WithIndexerAddress specifies the gRPC address to connect to the Indexer service.
//
// Use this to direct the client to perform Indexer service calls to a different address from the default.
// Matcher service calls will be directed to the default address unless WithMatcherAdress is specified.  
```

Also, maybe we can add a comment to `WithAddress`:

```
// WithIndexerAddress specifies the gRPC address to connect.
//
// The address will be used for both Indexer service and Matcher service calls. To use different
// addresses for each service, use With{Indexer,Matcher}Address.
```

**Review by jvdm — APPROVED** [review]
I would write those comments slightly differently, but I am nitpicking and things look good the way the are now if you want to move forward.

**Pull request merged**